src/Internal/BalanceTx/BalanceTx.purs

```diff
@@ -275,14 +282,28 @@ balanceTxWithConstraints transaction extraUtxos constraintsBuilder = do
 setTransactionCollateral :: Address -> Transaction -> BalanceTxM Transaction
 setTransactionCollateral changeAddr transaction = do
   nonSpendableSet <- asksConstraints _nonSpendableInputs
-  collateral <- do
-    rawCollateral <- liftEitherContract $ note CouldNotGetCollateral <$>
-      Wallet.getWalletCollateral
-    -- filter out UTxOs that are set as non-spendable in the balancer constraints
-    let
-      isSpendable = not <<< flip Set.member nonSpendableSet <<< _.input <<<
-        unwrap
-    pure $ Array.filter isSpendable rawCollateral
+  mbCollateralUtxos <- asksConstraints _collateralUtxos
+  -- We must filter out UTxOs that are set as non-spendable in the balancer
+  -- constraints
+  let
+    isSpendable = not <<< flip Set.member nonSpendableSet
+  collateral <- case mbCollateralUtxos of
+    -- if no collateral utxos are specified, use the wallet
+    Nothing -> Array.filter (isSpendable <<< _.input <<< unwrap) <$> do
+      liftEitherContract $ note CouldNotGetCollateral <$>
+        Wallet.getWalletCollateral
+    -- otherwise, get all the utxos, filter out unspendable, and select
+    -- collateral using internal algo, that is also used in KeyWallet
+    Just utxoMap -> do
```

Comment:
Also thinking, maybe it would make sense to add some error message in case the `utxoMap` is empty since the Transaction will fail by default because it won't have any UTXO to select as collateral?

Reply:
Btw. what would happen if I would set the `mustUseCollateralUtxos` to all the UTXOs in user's wallet? 🤔

Reply:
Only the required amount will be chosen

Reply:
A transaction should fail with a descriptive error message if it contains phase-2 scripts and the provided collateral candidate utxo map is empty.

Reply:
I added a runtime warning so that it's clear from the logs